SSA/Projects/InstCombine/ForLean.lean

```diff
@@ -574,6 +574,61 @@ theorem msb_sshiftRight' {x y: BitVec w} :
     (x.sshiftRight' y).msb = x.msb := by
   simp [BitVec.sshiftRight', BitVec.sshiftRight_msb_eq_msb]
 
+theorem getLsbD_sub {i : Nat} {i_lt : i < w} {x y : BitVec w} :
+    (x - y).getLsbD i =
+      (x.getLsbD i ^^ ((~~~y + 1).getLsbD i ^^ carry i x (~~~y + 1) false)) := by
+  rw [BitVec.sub_eq_add_neg, BitVec.neg_eq_not_add]
+  rw [getLsbD_add]
+  omega
+
+theorem getLsbD_neg {i : Nat} {i_lt : i < w} {x : BitVec w} :
+    getLsbD (~~~x) i = (getLsbD x i).not := by
+  simp
+  omega
+
+theorem getMsbD_add {i : Nat} {i_lt : i < w} {x y : BitVec w} :
+    getMsbD (x + y) i =
+      Bool.xor (getMsbD x i) (Bool.xor (getMsbD y i) (carry (w - 1 - i) x y false)) := by
+  simp only [getMsbD]
+  by_cases h₀ : i < w
+  · simp only [h₀, decide_True, Bool.true_and]
+    rw [BitVec.getLsbD_add]
+    omega
+  · simp [h₀]
+    omega
+
+theorem getMsbD_sub {i : Nat} {i_lt : i < w} {x y : BitVec w} :
+    (x - y).getMsbD i =
+      (x.getMsbD i ^^ ((~~~y + 1).getMsbD i ^^ carry (w - 1 - i) x (~~~y + 1) false)) := by
+  rw [BitVec.sub_eq_add_neg, BitVec.neg_eq_not_add]
+  rw [getMsbD_add]
+  omega
+
+theorem getMsbD_neg {i : Nat} {i_lt : i < w} {x : BitVec w} :
+    getMsbD (~~~x) i = (getMsbD x i).not := by
+  simp
+  omega
+
+theorem msb_add {w : Nat} {x y: BitVec w} :
+    (x + y).msb =
+      Bool.xor (getMsbD x 0) (Bool.xor (getMsbD y 0) (carry (w - 1) x y false)) := by
+  simp only [BitVec.msb, BitVec.getMsbD]
+  by_cases h₀ : w ≤ 0
+  · simp [h₀, show w = 0 by omega]
+  · simp [h₀, getLsbD_add, show w > 0 by omega]
+
+theorem msb_sub {x y: BitVec w} :
+    (x - y).msb
+      = (x.getMsbD 0 ^^ ((~~~y + 1#w).getMsbD 0 ^^ carry (w - 1 - 0) x (~~~y + 1#w) false)) := by
+    rw [BitVec.sub_eq_add_neg, BitVec.neg_eq_not_add,
+      msb_add]
+    rfl
```

Comment:
```suggestion
  simp [sub_eq_add_neg, BitVec.neg_eq_not_add, msb_add]
```

This can be shorter. Also, be careful with the indentation. Tactis use two whitespaces, not four as in this example.